### PR TITLE
strutils.countWhile: counts while a predicate is true, starting from either end

### PR DESCRIPTION
**Pull request by timotheecour**
* needed for subsequent PR's
* it's a useful and versatile proc that generalizes other procs

### CODE REVIEW

File: lib/pure/strutils.nim [hunk 1]
```diff
@@ -339,6 +339,25 @@ proc strip*(s: string, leading = true, trailing = true,
     while last >= 0 and s[last] in chars: dec(last)
   result = substr(s, first, last)
 
+proc countWhile*(str: string, pred : proc(a:char):bool, prefix = true): int =
```
Comment (Araq) [inline lib/pure/strutils.nim]: Nitpick: Watch the spacing around your colons.

Comment (Araq) [reply lib/pure/strutils.nim]: ``prefix`` should instead be ``countdown=false``.

Comment (Araq) [reply lib/pure/strutils.nim]: And I think ``s`` instead of ``str`` is more common in the stdlib.

File: lib/pure/strutils.nim [hunk 2]
```diff
@@ -339,6 +339,25 @@ proc strip*(s: string, leading = true, trailing = true,
...
+  ## returns the number of elements in ``str`` that satisfy ``pred`` predicate
+  ## without interruption, starting from beginning (when ``prefix`` is true)
+  ## or end (when ``prefix`` is false)
+  runnableExamples:
+    import sugar
+    doAssert countWhile("//abc", a=>a == '/') == 2
+    doAssert countWhile("abc//", a=>a == '/', prefix = false) == 2
+    doAssert countWhile("abcDEF", isLowerAscii) == 3
+
+  let lenS = len(str)
```
Comment (Araq) [inline lib/pure/strutils.nim]: This code looks better without the helper ``lenS``, use ``s.len`` instead.

File: lib/pure/strutils.nim [hunk 3]
```diff
@@ -339,6 +339,25 @@ proc strip*(s: string, leading = true, trailing = true,
...
+  if prefix:
+    for i in 0..<lenS:
+      if not pred(str[i]): return i
+  else:
+    for i in 0..<lenS:
```
Comment (Araq) [inline lib/pure/strutils.nim]: Use a ``countdown`` loop here instead.

File: lib/pure/strutils.nim (hunk 1 above)
Comment (Araq) [reply lib/pure/strutils.nim]: And I think the signature should be this, there is little you can do in this callback that receives a single ``char``...

```nim

proc countWhile*(s: string; x: set[char]; countdown = false): int
```

Or maybe 

```nim

template countWhileIt*(s: string; e: untyped; countdown = false): int
```

**Comment by dom96** [comment]
I believe I already explained why I don't think this function should make it into the stdlib.

Just to reiterate:

* It's incredibly specific thus unlikely to be used by many people.
* It can be easily emulated using functions defined in ``parseutils`` or using a simple ``for`` loop.
* I've searched a little and don't see a similar function in the stdlib of other languages. Most functional programming languages offer a ``filter`` function which would emulate the functionality of this procedure very well.

Nim actually already has a [``filter``](https://nim-lang.org/docs/sequtils.html#filter,openArray[T],proc(T)) function but it only supports ``openarray``s. A nice improvement would be to implement ``filter`` for strings.

**Comment by dom96** [comment]
> parseWhile allocates; countWhile doesn't

Okay, but couldn't you use ``skipUntil`` instead?

```nim
import parseutils

doAssert(skipUntil("abcDEF", {'A' .. 'Z'}) == 3)
```

> a set[char] is just not flexible enough, eg, there's no efficient or simple way to write countWhile("abcDEF", isLowerAscii) with parseWhile (would require precomputing a set of valid chars, which is both inefficient and unwieldy)

This is not inefficient nor unwieldy: ``{'a' .. 'z'}``.

> I need to allow going from either end (parseWhile doesn't, nor does anything in parseutils)

That is a problem. Reverse the string? :)

If that's too much of a performance hit then you can implement "prefix" mode for the ``parseutils`` procs.

**Pull request closed without merging**